pyro/poutine/runtime.py

```diff
@@ -196,3 +153,53 @@ def am_i_wrapped():
     :returns: bool
     """
     return len(_PYRO_STACK) > 0
+
+
+def effectful(fn=None, type=None):
+    """
+    :param fn: function or callable that performs an effectful computation
+    :param str type: the type label of the operation, e.g. `"sample"`
+
+    Wrapper for calling :func:~`pyro.poutine.runtime.apply_stack` to apply any active effects.
+    """
+    if fn is None:
+        return functools.partial(effectful, type=type)
+
+    if getattr(fn, "__wrapped", None):
+        return fn
+    else:
+        assert type is not None, "must provide a type label for operation {}".format(fn)
+        assert type != "message", "cannot use 'message' as keyword"
+
+        def _fn(*args, **kwargs):
```

Comment:
1. Consider using `@functools.wraps(fn)`
2. Would it be reasonable to pull out `name` from `*args`? I.e. are all foreseeable statements named? If so, then we could define exactly `param = effectful(_PYRO_PARAM_STORE.get_param, type="param")`.

Reply:
> I.e. are all foreseeable statements named?

No, actually I'd expect most other statements to not be named.